### PR TITLE
add Granulate gMaestro Addon

### PR DESCRIPTION
**Pull request by marina-vrublevsky**
*Issue #, if available:*

*Description of changes:*
This PR continues https://github.com/aws-samples/cdk-eks-blueprints-patterns/pull/57

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -234,6 +234,8 @@ For `kubeflow` please see [Kubeflow documentation](docs/patterns/kubeflow.md).
 
 For `secure-ingress-cognito`  please see [Secure Ingress using Cognito Blueprint documentation](docs/patterns/secureingresscognito.md).
 
+For `GmaestroConstruct` the pattern relies on the following secret defined, please see [Gmaestro documentation](https://github.com/Granulate/gmaestro-eks-blueprints-addon/blob/master/README.md):
```
Comment (elamaran11) [inline README.md:237]: You need a separate `md` file for Granulate like the other patterns you see. The pattern should clear have a complete documentation on how to deploy this pattern with all details, screenshots, diagram etc like example Kubeflow, Instana etc. You can add these details like secrets there.

**Review by elamaran11 — COMMENTED** [review]
@marina-vrublevsky Nice work on the patterns update. documentation file is missing like other patterns. Also details like secret should be removed in home `readme.md` file but rather should point to documentation
Also please share us the secret to validate this pattern in relevant AWS channel or POCs so we can consume to validate the pattern.

File: bin/gmaestro.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env node
+import GmaestroConstruct from '../lib/gmaestro-construct';
+import { configureApp, errorHandler } from '../lib/common/construct-utils';
+
+const app = configureApp();
+
+new GmaestroConstruct().buildAsync(app, 'gmaestro').catch((error) => {
+    errorHandler(app, "Gmaestro is not setup due to missing secrets: " + error);
```
[the remark this answers is not in the document]
Comment (marina-vrublevsky) [reply bin/gmaestro.ts:8]: Looks like the rest of them to me. e.g: `bin/instana-operator.ts`
Am i missing anything?

**Comment by marina-vrublevsky** [comment]
@elamaran11 please rereview
> Also please share us the secret to validate this pattern in relevant AWS channel or POCs so we can consume to validate the pattern.

will be provided soon

File: bin/gmaestro.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (elamaran11) [reply bin/gmaestro.ts:8]: Looks good

File: docs/patterns/gmaestro.md [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+## Prerequisite 
+Before using gMaestro, you need to:
+1. [Sign up](https://app.granulate.io/gMaestroSignup) to the gMaestro platform
+2. Download a sample YAML file - After signing up to gMaestro, navigate to the [Deploy](https://app.granulate.io/deploy) on the left-hand menu, fill in the required fields and click on "Generate Config File" 
+
+![GmaestroGenerateConfigFile](images/gmaestro-generate-config-file.png)
+
+![GmaestroConfigFile](images/gmaestro-config-file.png)
+
+3. Create a secret (as a plaintext) in AWS Secrets Manager copy its value from the following place:
```
Comment (elamaran11) [inline docs/patterns/gmaestro.md:21]: We should have a better mechanism to show to customer on how create secrets using CLI commands and also how to pass them to the solution. You should be using context to pass on fields mentioned below. Please see [this](https://github.com/aws-samples/cdk-eks-blueprints-patterns/blob/main/docs/patterns/instana.md) pattern for an example.

File: docs/patterns/gmaestro.md [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+cd cdk-eks-blueprints-patterns
+```
+
+Update the patterns parameters:
+
+- `clientIdSecretName: string` - The secret name from the prerequisite
+
+- `clusterName: string` - Copy from the Deployment section `MAESTRO_SERVICE_NAME` value
+
+- `createNamespace: boolean` - If you want CDK to create the namespace for you
```
Comment (elamaran11) [inline docs/patterns/gmaestro.md:41]: We should create this in a new namespace, we should avoid default namespace usage and not a good practice to show in patterns for customers. You can use `createNamespace` field.

**Review by elamaran11 — COMMENTED** [review]
Thankyou for your work on this PR, appreciate it. Have couple of minor comments if you can fix. Once you fix these, please share access to SAAS for validation.

**Comment by elamaran11** [comment]
@marina-vrublevsky Please fix the above issues pointed out.

**Comment by elamaran11** [comment]
@marina-vrublevsky Please provide an update on this PR and Patterns Repo PR.

**Comment by elamaran11** [comment]
@freschri As discussed, please pick this PR for tech validation. @marina-vrublevsky Please share the secrets for your share platform to us in some secured manner (you can email us or let us know we can invite you to AWS Dev slack).

**Comment by marina-vrublevsky** [comment]
@elamaran11 where can i email the secret?

**Comment by elamaran11** [comment]
> @elamaran11 where can i email the secret?

@marina-vrublevsky Share me your email, i will invite you to a slack workspace where you can share the secrets directly.

**Comment by elamaran11** [comment]
> @elamaran11 [marina.vrublevsky@granulate.io](mailto:marina.vrublevsky@granulate.io)

Hope you received the invite from Riccardo. Please share the secrets with him and myself in Slack.

**Review by freschri — APPROVED** [review]
approving following discussion on other channel and following amendments

File: docs/patterns/gmaestro.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+This pattern deploys the following resources:
+- Creates a single EKS cluster that includes a managed node group
+- Deploys supporting add-ons: ClusterAutoScaler and MetricsServer
+- Deploys a single granulate-gmaestro deployment with a single pod on the EKS cluster
+
+
+## Prerequisites
+Before using gMaestro, you need to:
+1. [Sign up](https://app.granulate.io/gMaestroSignup) to the gMaestro platform
+2. Download a sample YAML file - After signing up to gMaestro, navigate to the [Deploy](https://app.granulate.io/deploy) on the left-hand menu, fill in the required fields and click on "Generate Config File"
```
Comment (elamaran11) [inline docs/patterns/gmaestro.md:16]: same feedback as addon. Either attach a download hyperlink or add the same below. Screenshot wont help people to code this sample.

File: docs/patterns/gmaestro.md [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+![GmaestroGenerateConfigFile](images/gmaestro-generate-config-file.png)
+
+![GmaestroConfigFile](images/gmaestro-config-file.png)
+
+3. Create a secret (as a plaintext, not key/value) in AWS Secrets Manager:
+    ```bash
+    export MAESTRO_CLIENT_ID="<MAESTRO_CLIENT_ID value from the deployment section in the downloaded config file>"
+    export MAESTRO_SECRET_NAME="<MAESTRO_SECRET_NAME your preferred secret name>"
+    aws secretsmanager create-secret --name <MAESTRO_SECRET_NAME> \
```
Comment (elamaran11) [inline docs/patterns/gmaestro.md:26]: You are missing `region` parameter!

File: docs/patterns/gmaestro.md [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+```bash
+$ kubectl get pods -A | grep granulate-maestro
+
+NAMESPACE     NAME                                 READY   STATUS    RESTARTS   AGE
+default       granulate-maestro-6947dc87bc-k5nfc   1/1     Running   0          11m
+```
+
+After a few seconds, you will gain full visibility into your K8s cluster objects.
+The first rightsizing recommendations may take up to 5 minutes to load.
+
```
Comment (elamaran11) [inline docs/patterns/gmaestro.md:74]: Please add a couple of screenshot to show gMaestro in action with visibility and how you can right size...

File: docs/patterns/gmaestro.md [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+To clean up your EKS Blueprints, run the following commands:
+
+```sh
+make pattern gmaestro destroy
+```
+
+## Support
+
+If you have questions about Gmaestro, catch us [on Slack](https://granulatecommunity.slack.com/archives/C03RK0HN2TU)!
```
Comment (elamaran11) [inline docs/patterns/gmaestro.md:85]: This is very nice!

File: package.json [hunk 9]
```diff
@@ -39,7 +39,8 @@
     "@aws-sdk/client-eks": "^3.316.0",
     "eks-blueprints-cdk-kubeflow-ext": "0.1.9",
     "source-map-support": "^0.5.21",
-    "@instana/aws-eks-blueprint-addon": "^1.0.4"
+    "@instana/aws-eks-blueprint-addon": "^1.0.4",
+    "@granulate/gmaestro-eks-blueprints-addon": "^1.0.16"
   },
   "overrides": {
     "@aws-quickstart/eks-blueprints": "1.10.1",
```
Comment (elamaran11) [inline package.json:46]: Please reference the latest blueprints version in both places in `package.json`

File: README.md [hunk 10]
```diff
@@ -361,6 +361,9 @@ For `kubeflow` please see [Kubeflow documentation](docs/patterns/kubeflow.md).
 
 For `secure-ingress-cognito`  please see [Secure Ingress using Cognito Blueprint documentation](docs/patterns/secureingresscognito.md).
 
+For `GmaestroConstruct` the pattern relies on the following secret defined:
```
Comment (elamaran11) [inline README.md:364]: In this file `make list` command should be refreshed to show all patterns.

**Review by elamaran11 — CHANGES_REQUESTED** [review]
@marina-vrublevsky Nice work on the pattern, almost there. Couple of comments to close the loop with the PR to close all gaps. Also `mkdocs.yml` file update is missing, we need this so that website publishing is also taken care after merge.

File: package.json (hunk 9 above)
Comment (marina-vrublevsky) [reply package.json:46]: what is the second place?

**Review by elamaran11 — APPROVED** [review]
LGTM

**Comment by elamaran11** [comment]
@freschri Can you confirm in this PR, you were able to complete the test pattern with provided credentials from the Partner and validation worked fine?

**Comment by elamaran11** [comment]
@marina-vrublevsky I see some GH action failures

File: package.json [hunk 11]
```diff
@@ -39,10 +39,11 @@
     "@aws-sdk/client-eks": "^3.316.0",
     "eks-blueprints-cdk-kubeflow-ext": "0.1.9",
     "source-map-support": "^0.5.21",
-    "@instana/aws-eks-blueprint-addon": "^1.0.4"
+    "@instana/aws-eks-blueprint-addon": "^1.0.4",
+    "@granulate/gmaestro-eks-blueprints-addon": "^1.0.16"
   },
   "overrides": {
-    "@aws-quickstart/eks-blueprints": "1.10.1",
+    "@aws-quickstart/eks-blueprints": "1.11.0",
```
Comment (elamaran11) [inline package.json:46]: Change this to latest patch version

File: package.json [hunk 12]
```diff
@@ -24,7 +24,7 @@
     "typescript": "^5.0.4"
   },
   "dependencies": {
-    "@aws-quickstart/eks-blueprints": "1.10.1",
+    "@aws-quickstart/eks-blueprints": "1.11.0",
```
Comment (elamaran11) [inline package.json:27]: Change this to latest patch version

File: lib/gmaestro-construct/index.ts [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+        } as gmaestroAddOn.GmaestroAddOnProps;
+
+        const addOns: Array<blueprints.ClusterAddOn> = [
+            new blueprints.MetricsServerAddOn(),
+            new blueprints.addons.ClusterAutoScalerAddOn(),
+            new gmaestroAddOn.GmaestroAddOn(gmaestroAddOnProps)
+        ];
+        blueprints.EksBlueprint.builder()
+            .account(process.env.CDK_DEFAULT_ACCOUNT!)
+            .region(process.env.CDK_DEFAULT_REGION)
```
Comment (elamaran11) [inline lib/gmaestro-construct/index.ts:37]: .version('auto')

**Review by elamaran11 — CHANGES_REQUESTED** [review]
@marina-vrublevsky See this comment for fixing the GH Action issue.

**Comment by elamaran11** [comment]
@marina-vrublevsky Apologies, looks like there is some breaking changes to latest version. Can. you go back to previous version of blueprints and aws-cdk and remove `.version('auto')`. We can merge this PR and then i can work on fixing the breaking change with latest version.

**Comment by marina-vrublevsky** [comment]
> Can. you go back to previous version of blueprints 

which version? @elamaran11

**Comment by marina-vrublevsky** [comment]
@elamaran11 reverted to the version from main branch

**Comment by elamaran11** [comment]
> @elamaran11 reverted to the version from main branch
Blueprints version. I saw you had handled it.

**Review by elamaran11 — APPROVED** [review]
LGTM

**Pull request merged**